src/libs/vtkh/rendering/ConnectivityTracer.cpp

```diff
@@ -737,75 +737,91 @@ class IntegrateEmission : public vtkm::worklet::WorkletMapField
-
     vtkm::Id rayOffset = NumBins * rayIndex;
     vtkm::Id cellOffset = NumBins * currentCell;
-    for (vtkm::Int32 i = 0; i < NumBins; ++i)
-    {
-      const int cellOffsetI = cellOffset + i;
-      BOUNDS_CHECK(absorptionData, cellOffset + i);
-      FloatType absorb = static_cast<FloatType>(absorptionData.Get(cellOffset + i));
-      BOUNDS_CHECK(emissionData, cellOffset + i);
-      FloatType emission = static_cast<FloatType>(emissionData.Get(cellOffset + i));
 
-      absorb *= UnitScalar;
-      emission *= UnitScalar;
-
-      FloatType tmp = vtkm::Exp(-absorb * segmentLength);
-
-      //
-      // Traditionally, we would only keep track of a single intensity value per ray
-      // per bin and we would integrate from the beginning to end of the ray. In a
-      // distributed memory setting, we would move cell data around so that the
-      // entire ray could be traced, but in situ, moving that much cell data around
-      // could blow memory. Here we are keeping track of two values. Total absorption
-      // through this contiguous segment of the mesh, and the amount of emitted energy
-      // that makes it out of this mesh segment. If this is really run on a single node,
-      // we can get the final energy value by multiplying the background intensity by
-      // the total absorption of the mesh segment and add in the amount of emitted
-      // energy that escapes.
-      //
-      const int rayOffsetI = rayOffset + i;
-      BOUNDS_CHECK(absorptionBins, rayOffsetI);
-      FloatType absorbIntensity = static_cast<FloatType>(absorptionBins.Get(rayOffsetI));
-      BOUNDS_CHECK(emissionBins, rayOffsetI);
-      FloatType emissionIntensity = static_cast<FloatType>(emissionBins.Get(rayOffsetI));
-      BOUNDS_CHECK(opticalDepthBins, rayOffsetI);
-      FloatType opticalDepth = static_cast<FloatType>(opticalDepthBins.Get(rayOffsetI));
-
-      absorptionBins.Set(rayOffsetI, absorbIntensity * tmp);
-
-      if (DivideEmisByAbsorb)
+    //
+    // Traditionally, we would only keep track of a single intensity value per ray
+    // per bin and we would integrate from the beginning to end of the ray. In a
+    // distributed memory setting, we would move cell data around so that the
+    // entire ray could be traced, but in situ, moving that much cell data around
+    // could blow memory. Here we are keeping track of two values. Total absorption
+    // through this contiguous segment of the mesh, and the amount of emitted energy
+    // that makes it out of this mesh segment. If this is really run on a single node,
+    // we can get the final energy value by multiplying the background intensity by
+    // the total absorption of the mesh segment and add in the amount of emitted
+    // energy that escapes.
+    //
+
+    // NumBins can potentially be a very large number, so the loops are duplicated
+    // like this to avoid checking if DivideEmisByAbsorb == true each iteration.
+
+    if (DivideEmisByAbsorb)
```

Comment:
glad you pulled this out to the top level - there may be a way to refactor such that we don't have so much duplicated code in the two cases.